File: src/main/java/pl/training/module7/breakresponsibilities/after/DeploymentMetricsCalculator.java
package pl.training.module7.breakresponsibilities.after;

import java.util.List;
import java.util.Objects;

import pl.training.module7.breakresponsibilities.DeploymentMetrics;
import pl.training.module7.breakresponsibilities.DeploymentSample;

public final class DeploymentMetricsCalculator {
    public DeploymentMetrics calculate(List<DeploymentSample> samples) {
        Objects.requireNonNull(samples, "samples");

        int deployments = 0;
        int failures = 0;
        long totalLeadTimeMinutes = 0;

        for (DeploymentSample sample : samples) {
            Objects.requireNonNull(sample, "sample");
            deployments++;
            if (!sample.successful()) {
                failures++;
            }
            totalLeadTimeMinutes = Math.addExact(
                    totalLeadTimeMinutes, sample.leadTimeMinutes());
        }

        long averageLeadTimeMinutes = deployments == 0
                ? 0
                : totalLeadTimeMinutes / deployments;

        return new DeploymentMetrics(
                deployments, failures, averageLeadTimeMinutes);
    }
}

File: typescript/src/module7/breakresponsibilities/after/DeploymentMetricsCalculator.ts
import { requireNonNull } from '../../../shared/requireNonNull.js';
import { addExact } from '../../ExactMath.js';
import { DeploymentMetrics } from '../DeploymentMetrics.js';
import type { DeploymentSample } from '../DeploymentSample.js';

export class DeploymentMetricsCalculator {
  calculate(samples: readonly DeploymentSample[]): DeploymentMetrics {
    requireNonNull(samples, 'samples');

    let deployments = 0;
    let failures = 0;
    let totalLeadTimeMinutes = 0;

    for (const sample of samples) {
      requireNonNull(sample, 'sample');
      deployments++;
      if (!sample.successful) {
        failures++;
      }
      totalLeadTimeMinutes = addExact(
        totalLeadTimeMinutes, sample.leadTimeMinutes);
    }

    const averageLeadTimeMinutes = deployments === 0
      ? 0
      : Math.trunc(totalLeadTimeMinutes / deployments);

    return new DeploymentMetrics(
      deployments, failures, averageLeadTimeMinutes);
  }
}
